crates/assembly/src/linker/resolver/symbol_resolver.rs

```diff
@@ -0,0 +1,731 @@
+            let (ns, subpath) = path.split_first().unwrap();
+            log::debug!(target: "name-resolver::path", "resolving path as '{subpath}' relative to '{ns}'");
+            // Check if the first component of the namespace was previously imported
+            match self.resolve_import(context, Span::new(path.span(), ns))? {
+                SymbolResolution::Exact { gid, path } => {
+                    if subpath.is_empty() {
+                        log::debug!(target: "name-resolver::path", "resolved '{ns}' to imported item '{path}'");
+                        Ok(SymbolResolution::Exact { gid, path })
+                    } else {
+                        log::error!(target: "name-resolver::path", "resolved '{ns}' to imported item '{path}'");
+                        Err(Box::new(SymbolResolutionError::InvalidSubPath {
+                            span: context.span,
+                            source_file: self
+                                .graph
+                                .source_manager
+                                .get(context.span.source_id())
+                                .ok(),
+                            relative_to: None,
+                        })
+                        .into())
+                    }
+                },
+                SymbolResolution::MastRoot(digest) => {
+                    if subpath.is_empty() {
+                        log::debug!(target: "name-resolver::path", "resolved '{ns}' to imported procedure '{digest}'");
+                        Ok(SymbolResolution::MastRoot(digest))
+                    } else {
+                        log::error!(target: "name-resolver::path", "resolved '{ns}' to imported procedure '{digest}'");
+                        Err(Box::new(SymbolResolutionError::InvalidSubPath {
+                            span: context.span,
+                            source_file: self
+                                .graph
+                                .source_manager
+                                .get(context.span.source_id())
+                                .ok(),
+                            relative_to: None,
+                        })
+                        .into())
+                    }
+                },
+                SymbolResolution::Module { id, path: module_path } => {
+                    log::debug!(target: "name-resolver::path", "resolved '{ns}' to imported module '{module_path}'");
+                    if subpath.is_empty() {
+                        return Ok(SymbolResolution::Module { id, path: module_path });
+                    }
+
+                    let span = path.span();
+                    let path = module_path.join(subpath);
+                    let context = SymbolResolutionContext {
+                        span: module_path.span(),
+                        module: id,
+                        kind: context.kind,
+                    };
+                    self.resolve_path(&context, Span::new(span, path.as_path()))
```

Comment:
Idle thought: could a user-provided program with a lot of components blow the recursion stack? The simple mitigation would be to early reject on deep module paths `mod_1::..::mod_N` for e.g. `N > 100`.